### PR TITLE
fix(ibmi): use CurrentSchema insteadof DSN

### PR DESCRIPTION
**Pull request by hancockm**
Co-authored-by: renovate[bot] <29139614+renovate[bot]@users.noreply.github.com>

## Pull Request Checklist

- [ ] Have you added new tests to prevent regressions?
- [x] If a documentation update is necessary, have you opened a PR to [the documentation repository](https://github.com/sequelize/website/)? 
- [ ] Did you update the typescript typings accordingly (if applicable)?
- [x] Does the description below contain a link to an existing issue (Closes #[issue]) or a description of the issue you are solving?
- [x] Does the name of your PR follow [our conventions](https://github.com/sequelize/sequelize/blob/main/CONTRIBUTING.md#6-commit-your-modifications)?

## Description Of Change

https://github.com/sequelize/sequelize/issues/15227

Updated the ibmi dialect constructor and connect method. I added the property "DSN" and mapped the IBMi Schema/Library to the "database" property.  There was also a small correction to the "describeTable" method by adding a valdiation to check to see if "tableName" parameter is an object.

## Todos

- [ ] 
- [ ] 
- [ ]

### CODE REVIEW

**Comment by WikiRik** [comment]
Thanks for this PR! Can you point this PR towards `main` instead?

Tagging @markdirish in this as well since he worked on the implementation before

**Comment by hancockm** [comment]
@WikiRik . Done

File: src/dialects/ibmi/connection-manager.js [hunk 1]
```diff
@@ -35,8 +35,18 @@ export class IBMiConnectionManager extends AbstractConnectionManager {
 
     // 'database' doesn't make sense in this context, but it is mapped here to
```
Comment (WikiRik) [inline src/dialects/ibmi/connection-manager.js:36]: This comment needs to be updated

File: src/dialects/ibmi/connection-manager.js [hunk 2]
```diff
@@ -35,8 +35,18 @@ export class IBMiConnectionManager extends AbstractConnectionManager {
...
     // DSN, which is a close fit
+    //  if (config.database) {
+    //    connectionKeywords.push(`DSN=${config.database}`);
+    //  }
+
```
Comment (WikiRik) [inline src/dialects/ibmi/connection-manager.js:41]:
```suggestion
```

File: src/dialects/ibmi/query-generator.js [hunk 3]
```diff
@@ -130,21 +130,27 @@ export class IBMiQueryGenerator extends AbstractQueryGenerator {
   }
 
   dropTableQuery(tableName, options) {
-    let table = tableName;
+
```
Comment (WikiRik) [inline src/dialects/ibmi/query-generator.js:133]:
```suggestion
```

**Review by WikiRik — COMMENTED** [review]
Few style comments, but I do not really know anything about the dialect internally so if @markdirish says it's good I will agree with him. Looking forward to see the documentation update as well

File: src/dialects/ibmi/connection-manager.js (hunk 1 above)
Comment (hancockm) [reply src/dialects/ibmi/connection-manager.js:36]: Yes, I agree.

File: src/dialects/ibmi/index.js [hunk 4]
```diff
@@ -39,6 +39,10 @@ export class IBMiDialect extends AbstractDialect {
   constructor(sequelize) {
     super();
     this.sequelize = sequelize;
+
+    // Attach DSN configuation to the Sequelize object
+    this.sequelize.config.DSN = sequelize.options.DSN;
```
Comment (ephys) [inline src/dialects/ibmi/index.js:44]: Instanciating the dialect shouldn't cause any side effect. If DSN is an option specific to ibmi, I think it should go in `dialectOptions` instead.

File: src/dialects/ibmi/query-generator.js [hunk 5]
```diff
@@ -130,21 +130,26 @@ export class IBMiQueryGenerator extends AbstractQueryGenerator {
...
-    if (typeof table === 'object') {
-      schema = table.schema || undefined;
-      table = table.table;
-    } else if (options.schema) {
-      schema = options.schema;
-    }
 
-    return `DROP TABLE IF EXISTS ${schema ? `"${schema}".` : ''}"${table}"`;
+     if (typeof tableName === 'object') {
+      schema = tableName.schema || options.schema || undefined;
```
Comment (ephys) [inline src/dialects/ibmi/query-generator.js:138]: no other dialect accepts specifying the schema as part of the option bag, it's always part of `tableName`. You can drop `options` here.

(Not an issue for `describeTableQuery`, but we'll eventually need to drop the second parameter there too to be consistent)

**Review by ephys — CHANGES_REQUESTED** [review]

File: src/dialects/ibmi/query-generator.js (hunk 5 above)
Comment (WikiRik) [reply src/dialects/ibmi/query-generator.js:137]: I've gone ahead and fixed this in #15235

File: src/dialects/ibmi/query-generator.js [hunk 6]
```diff
@@ -130,21 +130,25 @@ export class IBMiQueryGenerator extends AbstractQueryGenerator {
...
+      tableName = tableName.table;
     }
 
-    return `DROP TABLE IF EXISTS ${schema ? `"${schema}".` : ''}"${table}"`;
+    return `DROP TABLE IF EXISTS ${schema ? `"${schema}".` : ''}"${tableName}"`;
   }
 
   describeTableQuery(tableName, schema) {
 
+    if (typeof tableName === 'object') {
```
Comment (WikiRik) [inline src/dialects/ibmi/query-generator.js:146]: This is now also fixed in #15235

**Comment by markdirish** [comment]
Finally got around to taking a look at this. A few thoughts:

1. Are you sure this is working as you expect? You map database to a `currentSchema` connection string keyword, but I don't see that as a valid keyword in our documentation: https://www.ibm.com/docs/en/i/7.5?topic=details-connection-string-keywords.  The closest thing appears to be `DBQ`, which specifies the libraries to add/replace values in the server job's library list. You say  that this fixed issues you had with `sequelize-automate`, but I'm not sure how if that keyword isn't valid. It might just be ignoring it?

2. If DBQ is the same behavior you are going for, we need to think about how to pass in the value in database. Note that in the documentation above, there are four different examples that all have different behavior:

> If UID=myuser;DBQ=mylib,mylib2,mylib3 is specified, then CURRENT SCHEMA special register is set to MYLIB and the library list would be MYLIB, MYLIB2, and MYLIB3.
> If UID=myuser;DBQ=,mylib,mylib2,mylib3;NAM=1 is specified, then CURRENT SCHEMA special register is set to *LIBL and the library list would be MYLIB, MYLIB2, and MYLIB3.
> If UID=myuser;DBQ=,mylib,mylib2,mylib3;NAM=0 is specified, then CURRENT SCHEMA special register is set to MYUSER and the library list would be MYLIB, MYLIB2, and MYLIB3.
> If UID=myuser;DBQ=mylib,*USRLIBL,mylib2 is specified, then CURRENT SCHEMA special register is set to MYLIB and the library list would be MYLIB, <existing user portion of the library list>, and MYLIB2.

Will run this through the sequelize integration tests to see what comes out the other end.

**Comment by WikiRik** [comment]
Closing this due to inactivity and the comment above which seems to suggest this change is invalid

**Pull request closed without merging**